src/java/io/bazel/rulesscala/scalac/ScalaCInvoker.java

```diff
@@ -0,0 +1,325 @@
+
+      if (file.isDirectory()) { // if its a directory, create it
+        f.mkdirs();
+        continue;
+      }
+
+      File parent = f.getParentFile();
+      parent.mkdirs();
+      outputPaths.add(f);
+
+      InputStream is = jar.getInputStream(file); // get the input stream
+      FileOutputStream fos = new FileOutputStream(f);
+      while (is.available() > 0) {  // write contents of 'is' to 'fos'
+        fos.write(is.read());
+      }
+      fos.close();
+      is.close();
+    }
+    return outputPaths;
+  }
+
+  static <T> String[] appendToString(String[] init, List<T> rest) {
+    String[] tmp = new String[init.length + rest.size()];
+    System.arraycopy(init, 0, tmp, 0, init.length);
+    int baseIdx = init.length;
+    for(T t : rest) {
+      tmp[baseIdx] = t.toString();
+      baseIdx += 1;
+    }
+    return tmp;
+  }
+  public static String[] merge(String[]... arrays) {
+    int totalLength = 0;
+    for(String[] arr:arrays){
+      totalLength += arr.length;
+    }
+
+    String[] result = new String[totalLength];
+    int offset = 0;
+    for(String[] arr:arrays){
+      System.arraycopy(arr, 0, result, offset, arr.length);
+      offset += arr.length;
+    }
+    return result;
+  }
+
+  /**
+   * This is the reporter field for scalac, which we want to access
+   */
+  private static Field reporterField;
+  static {
+    try {
+      reporterField = Driver.class.getDeclaredField("reporter"); //NoSuchFieldException
+      reporterField.setAccessible(true);
+    }
+    catch (Exception ex) {
+      throw new RuntimeException("nope", ex);
+    }
+  }
+
+  private static void processRequest(List<String> args) throws Exception {
+    Path tmpPath = null;
+    try {
+      if (args.size() == 1 && args.get(0).startsWith("@")) {
+        args = Files.readAllLines(Paths.get(args.get(0).substring(1)), UTF_8);
+      }
+      CompileOptions ops = new CompileOptions(args);
+
+      Path outputPath = FileSystems.getDefault().getPath(ops.outputName);
+      tmpPath = Files.createTempDirectory(outputPath.getParent(), "tmp");
+      String[] constParams = {
+        "-classpath",
+        ops.classpath,
+        "-d",
+        tmpPath.toString()
+        };
+
+      String[] compilerArgs = merge(
+        ops.scalaOpts,
+        ops.pluginArgs,
+        constParams,
+        extractSourceJars(ops, outputPath.getParent()));
+
+      MainClass comp = new MainClass();
+      long start = System.currentTimeMillis();
+      comp.process(compilerArgs);
+      long stop = System.currentTimeMillis();
+      System.err.println("Compiler runtime: " + (stop - start) + "ms.");
+
+      ConsoleReporter reporter = (ConsoleReporter) reporterField.get(comp);
+
+      if (reporter.hasErrors()) {
+          reporter.printSummary();
+          reporter.flush();
+          throw new RuntimeException("Build failed");
+      } else {
+        /**
+         * See if there are java sources to compile
+         */
+        if (ops.javaFiles.length > 0) {
```

Comment:
should we move this to `compileJavaSources()` since it is kind of a distraction for an unusual case in the code.